README.md
# Sui Name Service Python SDK

Python Sui Name Service SDK Client - built by community with [pysui](https://github.com/FrankC01/pysui/)

## Quick Start

Install `suinspy`

`pip install suinspy`

`poetry add suinspy`

Using git support:

`pip install git+https://git@github.com/andreidev1/suinspy.git`

Configure `pysui` with your own data.

```py
from pysui.sui.sui_config import SuiConfig
from pysui.sui.sui_clients.sync_client import SuiClient

def cfg_user():
    """Config user"""
    cfg = SuiConfig.user_config(
        # Required
        rpc_url="https://fullnode.mainnet.sui.io:443/",
        # Must be a valid Sui keystring (i.e. 'key_type_flag | private_key_seed' )
        prv_keys=["AIUPxQvxM18QggDDdTO0D0OD6PNVvtet50072d1grIyl"],
        # Needed for subscribing
        ws_url="wss://fullnode.mainnet.sui.io:443/",
    )
    return cfg

cfg = cfg_user()
client = SuiClient(cfg)

```

Import `suinspy`
```py
from suinspy.client import SuiNsClient
```

Create an instance of SuinsClient and choose network type (`mainet`, `testnet` or `devnet`).

```py
suins = SuiNsClient(client, 'mainet')
```

Fetch a name object:
```py
suins.get_name_object("suins.sui")
```

Fetch a name object including the owner and avatar:

**_NOTE:_** `show_owner` and `show_avatar` arguments are optional.
```py
suins.get_name_object("suins.sui", show_owner=True, show_avatar=True)
```

Fetch a SuiAddress linked to a name:
```py
suins.get_address("suins.sui")
```

## Official SuiNS Resources

[Official SuiNS Website](https://suins.io/)

[Official SuiNS Discord](https://discord.gg/suinsdapp)

## Ask A Question

Join Our Community [discord](https://discord.gg/CUTen9zu5h)


CHANGELOG.md
## [0.2.1] - 2023-09-13

### Changed
- pysui from `0.32.0` to `0.35.0`

### Removed
- `send_request()`


## [0.2.0] - 2023-08-23

### Added

- main network compatibility ( [#6](https://github.com/andreidev1/suinspy/issues/6) )
- tests against main network ( [#7](https://github.com/andreidev1/suinspy/issues/7) )
- new interfaces
- `black` and `flake8` deps
- tox.ini

### Changed



### Fixed


### Removed


## [0.1.3] - 2023-08-23

### Added



### Changed

- pysui v0.29.0 to pysui v0.32.0

### Fixed


### Removed


## [0.1.2] - 2023-07-21

### Added



### Fixed

- poetry lock version

### Removed



## [0.1.1] - 2023-07-21

### Added

- pysui 0.30.0

### Fixed

- [bug](https://github.com/andreidev1/suinspy/issues/2) Fix `suinspy` package installation

### Removed

- pysui 0.27.1

## [0.1.0] - 2023-07-20

### Added

- CHANGELOG.md
- README.md
- Sui Name Service types
- Parser and Queries
- Upload package to PyPi

### Changed

- types/objects.py


### Removed

- utils/config.py
- pydantic dependency

tests/conftest.py
"""Configuration test file"""

import pytest

from pysui.sui.sui_config import SuiConfig
from pysui.sui.sui_clients.sync_client import SuiClient



def network_client():
    return dict(main=dict(rpc_url="https://fullnode.mainnet.sui.io:443/", ws_url="wss://fullnode.mainnet.sui.io:443/"),
                test=dict(rpc_url="https://fullnode.testnet.sui.io:443/", ws_url="wss://fullnode.testnet.sui.io:443/"),
                dev=dict(rpc_url="https://fullnode.devnet.sui.io:443/", ws_url="wss://fullnode.devnet.sui.io:443/")
                )
                

@pytest.fixture(scope='module')
def init_config():
    """Init config"""
    cfg = SuiConfig.user_config(
        # Required
        rpc_url=network_client()["test"]["rpc_url"],
        # Must be a valid Sui keystring (i.e. 'key_type_flag | private_key_seed' )
        prv_keys=["AIUPxQxvY18QggDDdTO0D0OD6PNVvtet50072d1grIyl"],
        # Needed for subscribing
        ws_url=network_client()["test"]["ws_url"]
    )
    return cfg


@pytest.fixture(scope='module')
def init_client(init_config):
    """Init client"""
    client = SuiClient(init_config)
    return client


tests/test_suinspy.py
from suinspy import __version__


def test_version():
    assert __version__ == '0.1.0'


tests/integration/test_suins_client.py
"""Test against test network and main network"""
from suinspy.client import SuiNsClient
from suinspy.type.objects import NameObject

def test_get_suins_contract_objects_testnet(init_client):

    stub_suins_client = SuiNsClient(
        client=init_client,
        network_type='testnet',         
        contract_objects=dict(
        packageId="0xa789edcd4419bd9b3296afe6be7e537f985bebd48a2793530267e746221a8e56",
        registry="0x6e7e165b58b6f18bf87c438865d9ffc7722562c351bdca7b10a190cf33976308",
        reverseRegistry="0xc20a41b333f7955bf22bd46846318c979e3530e14f5913956d2debe17135139e",
        suins="0xdf5a2bcc7dcd60e89a956d554e3392dc50df1b1bb40048663102cbda0160590c"
        )
    )

    suins = SuiNsClient(client=init_client, network_type='testnet')

    contract_objects = suins.get_suins_contract_objects()

    assert stub_suins_client.contract_objects == contract_objects


def test_get_name_object_testnet(init_client):

    stub_name_object = NameObject(
        expiration_timestamp_ms='1716366440521',
        nft_id="0x229738a3c32e744c3d3df6b7c7984fc09e60eb540cff3eefcb38a44d535ba8b8",
        target_address=None,
        id="0x8196d35f40c39ff44b52d9ce6858ae5eb87070405bad919d1d03425aa98ab5bc",
        owner="0x478a3aa22f112f33b5454a2ca865621878d4a493e6f7e7e678ebe61cbf0efdb5",
        content_hash="ipfs://QmaLFg4tQYansFpyRqmDfABdkUVy66dHtpnkH15v1LPzcY",
    ).__dict__

    suins_client = SuiNsClient(client=init_client, network_type='testnet')

    name_object = suins_client.get_name_object("suins.sui", show_owner=True, show_avatar=True)

    assert stub_name_object == name_object

def test_get_suins_contract_objects_mainet(init_client):

    stub_suins_client = SuiNsClient(
        client=init_client,
        network_type='mainet',         
        contract_objects=dict(
        packageId="0xd22b24490e0bae52676651b4f56660a5ff8022a2576e0089f79b3c88d44e08f0",
        registry="0xe64cd9db9f829c6cc405d9790bd71567ae07259855f4fba6f02c84f52298c106",
        reverseRegistry="0x2fd099e17a292d2bc541df474f9fafa595653848cbabb2d7a4656ec786a1969f",
        suins="0x6e0ddefc0ad98889c04bab9639e512c21766c5e6366f89e696956d9be6952871"
        )
    )

    suins = SuiNsClient(client=init_client, network_type='mainet')

    contract_objects = suins.get_suins_contract_objects()

    assert stub_suins_client.contract_objects == contract_objects


suinspy/client.py
"""Sui Name Service Python SDK"""
import requests

from pysui.sui.sui_types.address import SuiAddress
from pysui.sui.sui_clients.sync_client import SuiClient
from pysui.sui.sui_builders.get_builders import GetDynamicFieldObject, DynamicFields

from suinspy.type.objects import SuiNSContract, NameObject
from suinspy.utils.constants import GCS_URL, DEVNET_JSON_FILE, TESTNET_JSON_FILE
from suinspy.utils.parser import parse_registry_response
from suinspy.utils.queries import get_avatar, get_owner


class SuiNsClient:
    """Interact with SuiNS Package"""

    def __init__(
        self, client: SuiClient, network_type: str, contract_objects: dict = None
    ):
        self.client = client
        self.network_type = network_type
        self.contract_objects = contract_objects

    def get_suins_contract_objects(self) -> SuiNSContract:
        """Get sui name service contract objects IDs"""

        if self.client:
            if self.network_type == "testnet":

                contract_url = GCS_URL + TESTNET_JSON_FILE

                response = requests.get(contract_url)

                if response.status_code == 200:
                    self.contract_objects = response.json()

            if self.network_type == "devnet":

                contract_url = GCS_URL + DEVNET_JSON_FILE

                response = requests.get(contract_url)

                if response.status_code == 200:
                    self.contract_objects = response.json()

            if self.network_type == "mainet":
                self.contract_objects = dict(
                    packageId="0xd22b24490e0bae52676651b4f56660a5ff8022a2576e0089f79b3c88d44e08f0",
                    registry="0xe64cd9db9f829c6cc405d9790bd71567ae07259855f4fba6f02c84f52298c106",
                    reverseRegistry="0x2fd099e17a292d2bc541df474f9fafa595653848cbabb2d7a4656ec786a1969f",
                    suins="0x6e0ddefc0ad98889c04bab9639e512c21766c5e6366f89e696956d9be6952871",
                )

        return self.contract_objects

    def get_dynamic_field_object(
        self, parent_object_id: SuiAddress, key: None, type="0x1::string::String"
    ) -> DynamicFields:
        dynamic_field_object = GetDynamicFieldObject(
            parent_object_id, dict(type=type, value=key)
        )

        return dynamic_field_object

    def get_name_object(
        self, name: str, show_owner=False, show_avatar=False
    ) -> NameObject:
        """Fetch object details"""

        [top_level_domain, domain] = name.split(".")[::-1]

        self.get_suins_contract_objects()

        build_registry_response = self.get_dynamic_field_object(
            self.contract_objects["registry"],
            [top_level_domain, domain],
            f'{self.contract_objects["packageId"]}::domain::Domain',
        )

        registry_response = self.client.execute(build_registry_response)

        name_object = parse_registry_response(registry_response.__dict__)

        nft_id = name_object["nft_id"]

        if show_owner == True:
            owner = get_owner(self.client, nft_id)

            name_object.update(owner=owner)

        if show_avatar == True:
            avatar = get_avatar(self.client, nft_id)

            name_object.update(content_hash=avatar)

        return name_object

    def get_address(self, name: str) -> SuiAddress:
        """Get address of sui domain"""

        owner_address = self.get_name_object(name, show_owner=True)["owner"]

        return owner_address

    """
    def get_name(self, address: SuiAddress):
        
        self.get_suins_contract_objects()
        print(self.contract_objects)
        build_reverse_registry_response = self.get_dynamic_field_object(
            self.contract_objects["reverseRegistry"],
            address,
            'address'
            )
        
        print(build_reverse_registry_response.__dict__)
        reverse_registry_response = self.client.execute(build_reverse_registry_response)

        return reverse_registry_response
    """
